### PR TITLE
App uses SQLite database and executes raw SQL query

### LINKED ISSUE #199: App uses SQLite database and executes raw SQL query
[App uses SQLite Database and execute raw SQL query. Untrusted user input in raw SQL queries can cause SQL Injection. Also sensitive information should be encrypted and written to the database.](https://github.com/MobSF/owasp-mstg/blob/master/Document/0x04h-Testing-Code-Quality.md#injection-flaws-mstg-arch-2-and-mstg-platform-2)

Standards
```
CWE: CWE-89: Improper Neutralization of Special
Elements used in an SQL Command ('SQL Injection') 
OWASP Top 10: M7: Client Code Quality
```
Files
`com/onaio/steps/helper/DatabaseHelper.java `

### PR DESCRIPTION
**Pull request by owais-vd**
close #199

### CODE REVIEW

File: src/main/java/com/onaio/steps/helper/DatabaseHelper.java [hunk 1]
```diff
@@ -129,4 +126,10 @@ public boolean isColumnExist(SQLiteDatabase db, String tableName, String columnN
 
         return isExist;
     }
+
+    public void dropTable(SQLiteDatabase db, String... tableNames) {
+        for (String table : tableNames) {
+            db.execSQL(String.format("DROP TABLE IF EXISTS %s", table));
```
Comment (ekigamba) [inline src/main/java/com/onaio/steps/helper/DatabaseHelper.java:132]: This does not make it safe from SQL injection.

1. Evaluate the uses of `rawQuery` and `execSQL` in the repository and use parametirized queries eg, `execSQL(query, args)`, `bindString` and/or `compileStatement`. Figure out which ones are affected by user-input
2. Make the appropriate changes

Here is a Google article that explains more on preventing SQL injection https://sites.google.com/site/smsdproject/home/sql-injections/hands-on-lab-practice

**Review by ekigamba — CHANGES_REQUESTED** [review]

File: src/main/java/com/onaio/steps/helper/DatabaseHelper.java (hunk 1 above)
Comment (owais-vd) [reply src/main/java/com/onaio/steps/helper/DatabaseHelper.java:132]: tried this `execSQL("DROP TABLE IF EXISTS  %s", Household.TABLE_NAME);` but getting exception. Additionally, we don't have security implications on that query because the query is static not dynamic.

Comment (ekigamba) [reply src/main/java/com/onaio/steps/helper/DatabaseHelper.java:132]: @owais-vd Yes, that should not work. You should have checked the link provided which provides more details. I still think that you should tackled #1 else I'll have to tackle it since it needs to go in the final report

Comment (ekigamba) [reply src/main/java/com/onaio/steps/helper/DatabaseHelper.java:132]: @owais-vd Sorry, I meant tackle this

> Evaluate the uses of `rawQuery` and `execSQL` in the repository and use parametirized queries eg, `execSQL(query, args)`, `bindString` and/or `compileStatement`. Figure out which ones are affected by user-input

Comment (owais-vd) [reply src/main/java/com/onaio/steps/helper/DatabaseHelper.java:132]: I tried `rawQuery` and `exceSQL` but both methods return `SQLiteException` with `bindArgs` and only work on data selection queries. I've also checked the suggestion link and they are also concatenating the `String` value in `execSQL` method instead `bindArgs`.

Comment (ekigamba) [reply src/main/java/com/onaio/steps/helper/DatabaseHelper.java:132]: Cool, have you tackled the first bullet here https://github.com/onaio/steps-app/pull/208#discussion_r875673134

Comment (owais-vd) [reply src/main/java/com/onaio/steps/helper/DatabaseHelper.java:132]: yeah, i also tried to evaluate the query with `compileStatement` with string binding but also got the same exception.

Comment (ekigamba) [reply src/main/java/com/onaio/steps/helper/DatabaseHelper.java:132]: Ok, it seems there's a misunderstanding. Let me rephrase

1. Search for uses of `rawQuery` and `execSQL` in the repository
2. Figure out which queries are affected by user-input
3. Convert them to use parametirized queries as given in the examples. If you don't convert them, share justification details on them so that we can use this in the report to describe why we still have such queries and the security implication

@owais-vd Does this make more sense?

Comment (owais-vd) [reply src/main/java/com/onaio/steps/helper/DatabaseHelper.java:132]: cool, that make sense. in the repository we are using both methods `rawQuery` and `execSQL` for static query i.e not required user-input, and none of the methods are affected by user-input i.e no user-input required in any query which is handled by `rawQuery` and `execSQL`. User input is required only for data insertion and update but we are using `db.insert()` and `db.update()` instead of a raw query.

Comment (ekigamba) [reply src/main/java/com/onaio/steps/helper/DatabaseHelper.java:132]: Cool, you can add this to the issue and provide evidence if need be. I think that's the main issue and we can get further recommendations/direction on these

Comment (owais-vd) [reply src/main/java/com/onaio/steps/helper/DatabaseHelper.java:132]: https://github.com/onaio/steps-app/issues/199#issuecomment-1131642537

**Review by ekigamba — APPROVED** [review]

**Review by ukanga — CHANGES_REQUESTED** [review]
But this is using String.format() which is effectively the same thing as a string concatenation with the plus sign. I would expect the security tools to still complain.

This assumes that String.format() does sanity checks for SQL, it does not but I expect the sqlite lib does with
 db.execSQL("DROP TABLE IF EXISTS ?", table);

**Comment by ekigamba** [comment]
1. Yes, you are correct that it only concatenates. 
2. Owais and I tried this but it's not possible with the currently provided methods in the SDK. You can only bind params such as table fields and search params I have also tried other variations
3. The table name is a constant and not a user input and therefore the warning on the issue is not relevant for this specific query(s).
4. :man_facepalming:  I might have misunderstood your comment in that you understood our justification but concatenation using `%` might silence the tool. Reason I approved it even though it still doesn't sanitize the table name constant

**Review by ukanga — APPROVED** [review]
